pacman/model/partitioner_splitters/abstract_splitters/abstract_splitter_slice.py

```diff
@@ -0,0 +1,417 @@
+    def _ratio(numerator, denominator):
+        """ Get the ratio between two values, with special handling for when\
+            the denominator is zero.
+
+        :param int numerator:
+        :param int denominator:
+        :rtype: float
+        """
+        if denominator == 0:
+            return 0.0
+        return numerator / denominator
+
+    @classmethod
+    def _find_max_ratio(cls, required, available, plan_n_time_steps):
+        """ Find the max ratio between the resources.
+
+        :param ResourceContainer required: the resources used by the vertex
+        :param ResourceContainer available:
+            the max resources available from the machine
+        :param int plan_n_time_steps: number of time steps to plan for
+        :return: the largest ratio of resources
+        :rtype: float
+        :raise None: this method does not raise any known exceptions
+        """
+        cpu_ratio = cls._ratio(
+            required.cpu_cycles.get_value(),
+            available.cpu_cycles.get_value())
+        dtcm_ratio = cls._ratio(
+            required.dtcm.get_value(), available.dtcm.get_value())
+        sdram_ratio = cls._ratio(
+            required.sdram.get_total_sdram(plan_n_time_steps),
+            available.sdram.get_total_sdram(plan_n_time_steps))
+        return max((cpu_ratio, dtcm_ratio, sdram_ratio))
+
+    @overrides(AbstractSplitterCommon.create_machine_vertices)
+    def create_machine_vertices(self, resource_tracker, machine_graph):
+        slices_resources_map = self.__split(resource_tracker)
+        for vertex_slice in slices_resources_map:
+            machine_vertex = self.create_machine_vertex(
+                vertex_slice, slices_resources_map[vertex_slice],
+                self.MACHINE_LABEL.format(
+                    self._governed_app_vertex.label, vertex_slice.lo_atom,
+                    vertex_slice.hi_atom),
+                get_remaining_constraints(self._governed_app_vertex))
+            machine_graph.add_vertex(machine_vertex)
+        self._called = True
+        return True
+
+    @abstractmethod
+    def create_machine_vertex(
+            self, vertex_slice, resources, label, remaining_constraints):
+        """ creates a machine vertex
+
+        :param Slice vertex_slice: vertex slice
+        :param ResourceTracker resources: resources
+        :param str label: human readable label for machine vertex.
+        :param remaining_constraints: none partitioner constraints.
+        :type remaining_constraints: iterable [Constraint]
+        :return: machine vertex
+        :rtype: MachineVertex
+        """
+
+    @abstractmethod
+    def get_resources_used_by_atoms(self, vertex_slice):
+        """ gets the resources of a slice of atoms from a given app vertex.
+
+        :param vertex_slice: the slice to find the resources of.
+        :return: ResourceContainer.
+        """
+
+    def _estimate_slices(self):
+        """ estimates the slices for when not already been split.
```

Comment:
"Estimates the slices for when the governed_app_vertex has not already been split."

Reply:
fixed